### PR TITLE
Statistics

### PR DESCRIPTION
**Pull request by arthursw**
Add the export statistics feature.

Provides 4 filters to include / exclude studies and subjects names.

### CODE REVIEW

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java [hunk 1]
```diff
@@ -143,4 +150,29 @@ private Dataset updateDatasetValues(final Dataset datasetDb, final Dataset datas
 		return Utils.toList(repository.findByDatasetAcquisitionExaminationStudyId(studyId));
 	}
 
+	@Override
+	public List<Object[]> queryStatistics(String studyNameInRegExp, String studyNameOutRegExp, String subjectNameInRegExp, String subjectNameOutRegExp) throws Exception {
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java:154]: Seems weird to me for this method to be in a Service and not a Repository class.
And little indentation / "login" comment pb :) (maybe extract a constant / Class for procedure name)

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java [hunk 2]
```diff
@@ -554,4 +598,53 @@ public String getUnit() {
...
+			throw new RestServiceException(new ErrorModel(HttpStatus.NOT_FOUND.value(), "No result found.", e));
+		} catch (Exception e) {
+			throw new RestServiceException(
+					new ErrorModel(HttpStatus.UNPROCESSABLE_ENTITY.value(), "Error while querying the database.", e));
+		}
+		zipSingleFile(statisticsFile, zipFile);
+
+		byte[] data = Files.readAllBytes(zipFile.toPath());
+		ByteArrayResource resource = new ByteArrayResource(data);
+
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java:643]: I think you can delete statisticsFile here

File: docker-compose/database/3_add_statistics_procedure.sql [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+-- the Free Software Foundation, either version 3 of the License, or
+-- (at your option) any later version.
+-- 
+-- You should have received a copy of the GNU General Public License
+-- along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
+
+use datasets;
+
+delimiter //
+
```
Comment (jcomedouteau) [inline docker-compose/database/3_add_statistics_procedure.sql:16]: Have you tested to enter SQL injection like, or comments in the entries to check security ?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java [hunk 4]
```diff
@@ -486,6 +493,43 @@ private void zip(final String sourceDirPath, final String zipFilePath) throws IO
...
+	
+	/**
+	 * Zip a single file
+	 * 
+	 * @param sourceFile
+	 * @param zipFile
+	 * @throws IOException
+	 */
+	private void zipSingleFile(final File sourceFile, final File zipFile) throws IOException {
+
```
[the remark this answers is not in the document]
Comment (arthursw) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java:505]: I wanted a function to zip a single file, not a directory, that's why I created this function (zipSingleFile vs zip). I could have made a directory with my single file but I felt it's better this way.

Thanks for pointing to "try with ressource" doc!

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java [hunk 5]
```diff
@@ -554,4 +598,53 @@ public String getUnit() {
...
+			@ApiParam(value = "Study name including regular expression", required=false) @Valid
+			@RequestParam(value = "studyNameInRegExp", required = false) String studyNameInRegExp,
+			@ApiParam(value = "Study name excluding regular expression", required=false) @Valid
+			@RequestParam(value = "studyNameOutRegExp", required = false) String studyNameOutRegExp,
+			@ApiParam(value = "Subject name including regular expression", required=false) @Valid
+			@RequestParam(value = "subjectNameInRegExp", required = false) String subjectNameInRegExp,
+			@ApiParam(value = "Subject name excluding regular expression", required=false) @Valid
+			@RequestParam(value = "subjectNameOutRegExp", required = false) String subjectNameOutRegExp
+	) throws RestServiceException, IOException {
+		String tmpDir = System.getProperty(JAVA_IO_TMPDIR);
```
[the remark this answers is not in the document]
Comment (arthursw) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java:612]: Yep, I copied past the above code. Now that I updated to the latest version, I see and use the `getUserImportDir` function 👍

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java [hunk 6]
```diff
@@ -554,4 +598,53 @@ public String getUnit() {
...
+			FileOutputStream fos = new FileOutputStream(statisticsFile);
+			BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(fos));
+
+			for (Object[] or : results) {
+				List<String> strings = Arrays.stream(or).map(object -> Objects.toString(object, null)).collect(Collectors.toList());
+				bw.write(String.join("\t", strings));
+				bw.newLine();
+			}
+		 
+			bw.close();
```
[the remark this answers is not in the document]
Comment (arthursw) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/DatasetApiController.java:631]: 👍

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetService.java [hunk 7]
```diff
@@ -102,4 +102,12 @@
 	@PostAuthorize("hasRole('ADMIN') or @datasetSecurityService.checkDatasetPage(returnObject, 'CAN_SEE_ALL')")
 	public List<Dataset> findByStudyId(Long studyId);
 
+	/**
+	 * Get database statistics
+	 * 
+	 * @return statistics
+	 */
+	@PreAuthorize("hasRole('ADMIN')")
```
[the remark this answers is not in the document]
Comment (arthursw) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetService.java:110]: Well, I don't really know ; maybe experts too, I will ask Michael about that.

File: docker-compose/database/3_add_statistics_procedure.sql (hunk 3 above)
Comment (arthursw) [reply docker-compose/database/3_add_statistics_procedure.sql:16]: J'ai checké, 
```StoredProcedureQuery query [...]
query.setParameter(1, studyNameInRegExp);```
Permet de se prémunir des injections SQL.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java (hunk 1 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java:154]: as the web GUI is used, and normally the controller goes to the service, I think this is fine for me

**Review by michaelkain — APPROVED** [review]

**Pull request merged**